### ISSUE TITLE
GenotypeGVCFs memory/Xmx when reading GenomicsDB workspace as input?

### ISSUE THREAD

**Issue by bbimber** [broadinstitute/gatk#7542]
Hello,

When using GATK GenomicsDbImport, I understand that both the java layer and C layer use memory and therefore we need to allow a buffer for the genomics DB code. We're getting odd behavior from GenotypeGVCFs memory. We run a command similar to:

```
java -Djava.io.tmpdir=<path> -Xmx384g -Xms384g -Xss2m -jar GenomeAnalysisTK4.jar GenotypeGVCFs -R <FASTA> --variant gendb:///.....gdb -O <path> --annotate-with-num-discovered-alleles -stand-call-conf 30 --max-alternate-alleles 12 --force-output-intervals <BED_FILE> -L 3:129881515-132852846 --only-output-calls-starting-in-intervals --genomicsdb-shared-posixfs-optimizations

```

We find the job is using ~466GB, not the 384 I'd expect it to be constrained by. Does using a genomics DB as input result in some other non-java code being used? I'm trying to understand why Xmx/Xmx isnt constraining it. 

Thanks for any help.

**Comment by droazen** [comment]
@bbimber The -Xmx option controls only the maximum size of the *Java* heap. It does not limit the size of the C heap used by the GenomicsDB library. This is why we always advise leaving extra memory available for C when using GenomicsDB by selecting an -Xmx value that is comfortably under the amount of physical memory available. Even though your GenotypeGVCFs command is not importing into a GenomicsDB, it is reading out of a GenomicsDB and performing an on-the-fly merge, which will cause the native library to use nontrivial amounts of memory.

Having said that, ~80 GB for the native heap does seem like a lot. Do you have lots of highly multi-allelic records in your callset? @nalinigans @mlathara Thoughts on this one?

**Comment by bbimber** [comment]
@droazen Yes, i get there's a difference in java vs. C. We have thinking about reserving memory for C specifically for GenomicsDbImport (which is obviously very C heavy), and we have been ignoring this for GenotypeGVCFs with a workspace as input. You're confirming that we need to give this thought whenever using code that involves a GenomicsDb workspace.

Yes, we probably do have a lot of highly multi-allelic records. this is ~2300 WGS datasets.

**Comment by lbergelson** [comment]
@bbimber Something about the way this is running seems strange to me.  We typically run GenotypeGvcfs with ~26GB of memory a so I'm surprised you're seeing it need an order of magnitude more.  The memory requirements should roughly scale with the number of samples rather than the number of variants since it processes it basically line by line.  Sites with high numbers or alleles can really explode the memory requirements though, I see you're setting the max alt alleles to 12 here, which is a good idea, so I'm not totally sure where the high memory overhead is coming from.

Surprisingly most of the work of genomicsDB is done when you READ from it rather than when importing it.  The import is mostly a write operation, but reading triggers all the expensive merging operations necessary to create a coherent view of all the samples.  I would expect reading from it to use as much or more memory than writing to it.  That said, 80gb is more than I would expect it to need.

**Comment by bbimber** [comment]
@lbergelson I completely agree. GenomicsDB has been a big step up from CombineGVCFs, but it's been a struggle. Here are some notable characteristics of our data, in case anything seems relevant:

- The core data is ~2K WGS (not WXS) rhesus macaque datasets. WXS tends to perform better.
- We imported them into a workspace using GenomicsDB. 
- We regularly append new batches. As the data has grown, we found we need to lower the batch size (maybe 50-100) to make GenomicsDbImport practically work. It's possible a --consolidate type step might help us; however, our earlier efforts to run this resulted in hung jobs.
- We execute these jobs scattered across a cluster that is using a lustre filesystem. We use the non-posix optimization flag.
- We include the unplaced contigs, so our genome is ~2,900 contigs.
- We currently ask GenotypeGVCFs to call sites genome-wide. I've been pondering whether we should mask repetitive regions. This might also have the effect of removing sites with incredibly high numbers of distinct alleles 

I'd appreciate any ideas you or the GATK team has.

**Comment by mlathara** [comment]
I don't really have any further insight here, but I do think the max-alternate-allele and max-genotype is the issue (as already stated). Fwiw, the default value for max-alternate-allele is `6`, so you're setting it higher than that. Of course, ploidy also plays a role in number of genotypes. 

The combination of high ploidy and higher max alternate allele setting can result in a ~10x increase in number of genotypes. Presumably, you need the higher max-alternate-allele setting so not sure we can do much to work around this. But an experiment to set that lower (or use the default), and/or to set the max-genotypes to less than `1024` would help confirm this.

**Comment by mlathara** [comment]
Btw - it is interesting (as in, unexpected, at least to me) that appending new data is slowing down as the workspace grows. Adding new data is sorta fire-and-forget in that it shouldn't care much about what already exists...offhand, I'm not sure why you're seeing a slowdown. You mentioned needing to use smaller batch sizes...were you otherwise seeing larger memory overheads than before? Or just general slower performance?

There is a new `--bypass-feature-reader` option in the latest release that should help with dramatically lowering memory usage, and potentially offering a slight speedup for imports. Might be worth a shot.

**Comment by bbimber** [comment]
OK. As a reference, how does GATK deal with max-alternate-alleles for normal human variant calling? Presumably really high alternate alleles would primarily happen in repetitive/index prone-regions? FWIW, When we execute GenotypeGVCFs, we run as ~1000 jobs where each takes an even chunk of the genome, by base pairs. 

Yes, I did see the bypass-feature-reader option, but we have jobs in-flight and I'm reluctant to change too many things as once. We will try this when possible though.

As far as number of batches imported: I would need to check, but I believe it's only ~5 batches with perhaps 50-100 samples/ea. So I guess it's not that many new batches in the scheme of things, but anecdotally we have noticed that with the last couple rounds of import we needed to reduce batch size to make it work (i.e. not get hung). It is conceivable there is some other factor that is causing that variable performance.

**Comment by nalinigans** [comment]
@bbimber, there was an issue with `max-alternate-alleles` getting passed to the GenomicsDB layer from `GenotypeGVCFs`. That has been fixed in this [branch](https://github.com/broadinstitute/gatk/tree/ng_genomicsdb_args) if you would like to try.  See this [GenomicsDB Issue](https://github.com/GenomicsDB/GenomicsDB/commit/3c7d1ead0110fec26d56ff85a5871d8df673504d). This will hopefully help with your memory issue.

On another note, there is a [performance/bug fix ](https://github.com/broadinstitute/gatk/pull/7520) while reading/writing GenomicDB bookkeeping artifacts in the latest release and in this branch as well. This may help with the memory issue while incrementally adding new batches and using the `--consolidate1 option with import.
